connections/multiselect2.md

```diff
@@ -0,0 +1,108 @@
+
+To determine the roles in the second handshake attempt, endpoints compare the SHA-256 hashes of their peer IDs. The peer with the numerically smaller hash value acts as a client in the second handshake attempt, the peer with the numerically larger hash value acts as a server.
+
+Since secio assign roles during the handshake, it is not possible to detect a Simultaneous Open in this case. Therefore, secio MUST NOT be used with Multiselect 2.0.
+
+### Stream Multiplexer Selection
+
+This section only applies if Multiselect 2 is run over a transport that is not natively multipexed. Transports that provide stream multiplexing on the transport layer (e.g. QUIC) don't need to do anything described in this section.
+
+Some handshake protocols (TLS 1.3, Noise) support sending of *Early Data*. 
+
+In Multiselect 2 the server makes use of Early Data by sending a list of stream multiplexers. This ensures that the client can choose a stream multiplexer as soon as the handshake completes (or fail the connection if it doesn't support any stream multiplexer offered by the server).
+
+When using TLS 1.3, the server can send Early Data after it receives the ClientHello. Early Data is encrypted, but at this point of the handshake the client's identity is not yet verified.
+While Noise in principle allows sending of unencrypted data, endpoints MUST NOT use this to send their list of stream multiplexers. An endpoint MAY send it as soon it is possible to send encrypted data, even if the peers' identity is not verified at that point.
+
+Note that this negotiation scheme allows peers to negotiate a "monoplexed" connection, i.e. a connection that doesn't use any stream multiplexer. Endpoints can offer support for monoplexed connections by offering the `/monoplex` stream multiplexer.
+
+**TODO**: Do we need to define a way to send an error code / error string? Or do we have something like that in libp2p already?
+
+![](handshake.png)
+
+Handshake protocols (or implementations of handshake protocols) that don't support sending of Early Data will have to run the stream multiplexer selection after the handshake completes.
+
+#### 0-RTT
+
+When using 0-RTT session resumption as offered by TLS 1.3 and Noise, the endpoints MUST remember the negotiated stream multiplexer used on the original connection. This ensures that the client can send application data in the first flight when resuming a connection.
+
+## Protocol Speficiation
+
+All messages are Protobuf messages using the `proto3` syntax. Every message is wrapped by the `Multiselect` message:
+
+```protobuf
+# Wraps every message
+message Multiselect {
+    oneof message {
+        Offer offer = 1;
+        Use use = 2;
+    }
+}
+```
```

Comment:
Nit: for readability I'd put the definition of `Offer` and `Use` here and not below.

Reply:
Yes, we need the definitions and protocol overview before the wire format.

Reply:
And can we have the entire schema dumped at once after the overview. It's difficult to reason when types definitions are split across prose.

Reply:
Agreed. I'll do another editorial pass over the document soon, and will apply this suggestion.

Reply:
I did another editorial pass over the document and moved the protobuf definitions to a separate section.